Copilot: update backend to use new /metrics API

henrikedegrd opened the pull request:
## Hey, I just made a Pull Request!

I have updated the backend to use the new /metrics API provided by Github since the old /usage API now is deprecated and not gives out new data since 30th of January.

The Backend stores all the new data into its own tables in order to build more visualisations going forward.

Currently the Backend converts the output of the database into the old format that the Frontend uses.
This is not optimal, but in order to not loose data (since the API only gives data for the last 28 days), I have opted for this solution for now. But it should be changed in the future since the convertion is not optimal.

I have tested this locally towards Organization, Enterprise*Tasks not tested, but should work since they are almost copy-paste.

Made a small text-change in the Frontend to a table that had the wrong title.

Once this is merged more work can be done in order to show more metrics in the frontend.

This is my first contribution, so I have probably missed something.

There might be some changes in here that first was added in #2835 since I used that as a base.

This addresses #2828 

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/community-plugins/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/community-plugins/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

Zarthus reviewed (approved):
Thank you - this looks a lot better and more thorough than my stab at it :)

henrikedegrd commented:
> We can't [easily] test this on our staging instance of backstage - we don't have the right access controls to modify files inplace from node_modules like the way of testing you suggested, unfortunately.

Found a way to test the changes pretty easily:
1. Clone the repo
2. Pack the plugins with 'yarn pack' (run that in each of the 3 packages (copilot, copilot-common and copilot-backend))
3. Copy the produced files (package.tgz) into a folder in your backstage project (local_packages in my case). Name them according example below.
4. Reference them in package.json (both in app and backend) like this:

backend/package.json:
```
"@backstage-community/plugin-copilot-backend": "file:../../local_packages/copilot-backend.tgz",
"@backstage-community/plugin-copilot-common": "file:../../local_packages/copilot-common.tgz",

```

app/package.json
```
"@backstage-community/plugin-copilot": "file:../../local_packages/copilot.tgz",
"@backstage-community/plugin-copilot-common": "file:../../local_packages/copilot-common.tgz",
```

awanlin commented on workspaces/copilot/plugins/copilot-backend/src/task/TaskManagement.ts:
```diff
@@ -54,9 +54,15 @@ export default class TaskManagement {
+            `[TaskManagement] Failed to process task: ${e.message}`,
+          );
+        } else {
+          this.options.logger.warn(
```
```suggestion
          this.options.logger.error(
```

These are both errors, right? I know it was just a warning before but this does seem like it should be changed.

henrikedegrd replied:
Yes, you are correct, they are errors and should be logged as errors.

awanlin commented on workspaces/copilot/plugins/copilot-common/src/index.ts:
```diff
@@ -23,6 +23,10 @@
+ * that was deprecated. At some point we should fully convert to the new CopilotMetrics way.
+ * Until the frontend is fully converted, we need to keep this type, since the new metrics converts its result to this type.
+ *
  * @public
```
```suggestion
 * @deprecated
```

Let's mark it as deprecated so we can signal the intent to remove it at a future point more easily 👍

awanlin commented on workspaces/copilot/plugins/copilot-common/src/index.ts:
```diff
@@ -71,6 +75,9 @@ export type MetricsType = 'enterprise' | 'organization';
+ * Until the frontend is fully converted to the new CopilotMetrics way,
+ * we need to keep this type, since the new metrics converts its result to this type.
  *
  * @public
```
```suggestion
* @deprecated
```

Same here

awanlin commented on workspaces/copilot/.changeset/mighty-pears-compete.md:
```diff
@@ -0,0 +1,8 @@
+---
+
+Backend updated for using the new /metrics API that Github Provides.
+Added new tables to the database to store all metrics provided by Github.
```
I think it would make sense to include a bit of detail here as to how it handles the older data. There's some logic in the router for this that's worth mentioning.

henrikedegrd replied:
Updated the changeset with some better details. Hope the details will make sense :)

awanlin reviewed (changes requested):
Thanks for running with this @henrikedegrd, left a few comments, nothing hugely structural though.

henrikedegrd commented:
Updated the PR with changes from comments and also fixed the /period-range endpoint in the backend.

I somehow missed that, it should also like /metrics endpoint take both old and new metrics into consideration when giving out the availible date-range.

awanlin commented on workspaces/copilot/.changeset/mighty-pears-compete.md:
```diff
@@ -0,0 +1,19 @@
+---
+'@backstage-community/plugin-copilot-backend': minor
+'@backstage-community/plugin-copilot-common': minor
+'@backstage-community/plugin-copilot': minor
```
The frontend changes are just removing a space and then some text changes. These don't seem directly related and also seem more like a patch. Can we remove them for a separate PR or create a separate changeset for them at `patch` level, please?

henrikedegrd replied:
I added a separate changeset, a bit new into this, so if it cant be 2 changesets in this PR i will remove and do a separate PR for the frontend-changes.

awanlin reviewed (changes requested):
Thanks for the quick follow up @henrikedegrd, left one more comment that got missed on my end the last time around.

awanlin reviewed (approved):
Awesome!

henrikedegrd commented:
I see that the CI gives error on the api-reports. But when i try to update them, it gives me a error because a deprecated type is still used in the api (it seems it also needs @public?)

awanlin commented:
> I see that the CI gives error on the api-reports. But when i try to update them, it gives me a error because a deprecated type is still used in the api (it seems it also needs @public?)

Ah, yes, that's correct, needs both, sorry about that :(

henrikedegrd commented:
> Ah, yes, that's correct, needs both, sorry about that :(

No worries, updated them now with success :)
Thanks for the review!

Will continue to update the plugin with more detailed metrics in the following weeks.

awanlin commented:
Hi @Zarthus, these changes are ready to merge but I'd like one more confirmation from you, please?

It would be awesome if you can test this in a working setup. You can do what @henrikedegrd suggested here: https://github.com/backstage/community-plugins/pull/2989#issuecomment-2687168099.
Another option is to try this: https://backstage.io/docs/tooling/local-dev/linking-local-packages#generating-temporary-patches

Zarthus commented:
@awanlin I'm on PTO next week (I did file an internal ticket for my teammates to do the testing in my absence but I'm not sure if they'll get time for it). Earliest I can give it a shot is probably Tuesday 11th.

dkadirbekovbeeline commented:
Hi everyone!
When can we apply this patch? We’re already experiencing the same issue in our project.

Zarthus commented:
@dkadirbekovbeeline everyone is experiencing this issue, I encourage following the testing instructions here to help accelerate:  https://github.com/backstage/community-plugins/pull/2989#issuecomment-2692387300

semihural-tomtom commented:
+1

awanlin commented:
Hi @henrikedegrd @Zarthus @dkadirbekovbeeline @semihural-tomtom, I'm going to merge this and release the updated version. I'm asking you all to please test this as soon as it's release just to confirm there are no further issues. Thanks!

Zarthus commented:
Been running on staging fine since Monday, deployed to prod, also fine. It works 👍🏻

awanlin commented:
Thanks for the follow up @Zarthus, appreciate it! 🚀

dkadirbekovbeeline commented:
Sorry for stupid question, but how can I update version of copilot plugin?

dkadirbekovbeeline commented:
I'm getting error: 
`[backend]: 2025-03-12T09:30:20.033Z copilot info [discoverEnterpriseMetrics] Found last day: undefined
[backend]: 2025-03-12T09:30:20.036Z copilot info [discoverEnterpriseMetrics] Found 27 new metrics to insert
[backend]: 2025-03-12T09:30:20.038Z copilot error [discoverEnterpriseMetrics] An error occurred while processing Github Copilot metrics: TypeError: Cannot read properties of undefined (reading 'map')`

henrikedegrd commented:
> I'm getting error: `[backend]: 2025-03-12T09:30:20.033Z copilot info [discoverEnterpriseMetrics] Found last day: undefined [backend]: 2025-03-12T09:30:20.036Z copilot info [discoverEnterpriseMetrics] Found 27 new metrics to insert [backend]: 2025-03-12T09:30:20.038Z copilot error [discoverEnterpriseMetrics] An error occurred while processing Github Copilot metrics: TypeError: Cannot read properties of undefined (reading 'map')`
> 
> `[backend]: 2025-03-12T09:30:20.072Z copilot info [discoverOrganizationMetrics] Fetched 27 metrics [backend]: 2025-03-12T09:30:20.075Z copilot info [discoverOrganizationMetrics] Found last day: undefined [backend]: 2025-03-12T09:30:20.077Z copilot info [discoverOrganizationMetrics] Found 27 new metrics to insert [backend]: 2025-03-12T09:30:20.078Z copilot error [discoverOrganizationMetrics] An error occurred while processing Github Copilot metrics: TypeError: Cannot read properties of undefined (reading 'map')`

Seems like the same problem here: #3182 
That resolved itself somehow, and I am unable to reproduce this error.
But as I wrote in the issue, I will try to make the error-messages a bit more informative.
